reactor-core/src/main/java/reactor/core/publisher/SinksSpecs.java

```diff
@@ -180,27 +168,38 @@ public void emitNext(T value) {
 	public final Emission tryEmitNext(T t) {
 		Objects.requireNonNull(t, "t is null in sink.next(t)");
 
-		Thread currentTread = Thread.currentThread();
-		if (WIP.get(this) == 0 && WIP.compareAndSet(this, 0, 1)) {
-			LOCKED_AT.lazySet(this, currentTread);
-		}
-		else {
-			if (LOCKED_AT.get(this) != currentTread) {
-				return Emission.FAIL_NON_SERIALIZED;
-			}
-			WIP.incrementAndGet(this);
+		Thread currentThread = Thread.currentThread();
+		if (!tryAcquire(currentThread)) {
+			return Emission.FAIL_NON_SERIALIZED;
 		}
 
 		try {
 			return sink.tryEmitNext(t);
 		}
 		finally {
 			if (WIP.decrementAndGet(this) == 0) {
-				LOCKED_AT.compareAndSet(this, currentTread, null);
+				LOCKED_AT.compareAndSet(this, currentThread, null);
 			}
 		}
 	}
 
+	private boolean tryAcquire(Thread currentThread) {
+		if (WIP.get(this) == 0 && WIP.compareAndSet(this, 0, 1)) {
+			// lazySet here, because:
+			// 1. initial state is `null`
+			// 2. `LOCKED_AT.get(this) != currentThread` from the next branch will either see null or an outdated old thread
+			// 3. The only possibility to make the condition pass is to read it from the current thread that already has the value cached
```

Comment:
```suggestion
			// lazySet in thread A here is ok because:
			// 1. initial state is `null`
			// 2. `LOCKED_AT.get(this) != currentThread` from a different thread B could see outdated null or an outdated old thread
			// 3. but that old thread cannot be B: since we're in thread B, it must have executed the compareAndSet which would have loaded the update from A
			// 4. Seeing `null` or `C` is equivalent from seeing `A` from the perspective of the condition (`!= currentThread` is still true in all three cases)
```